### Cardiac Arrhythmia Multy-Class Classification 

Analyze data and address missing data if there is any. 

Decide aboute a good evaluation strategy and justify your choice. 

Find the best parameters for the following classification models: 
- KNN classifcation 
- Logistic Regression
- Linear Supprt Vector Machine
- Kerenilzed Support Vector Machine
- Decision Tree
- Random Forest 

Then use different bagging and boosting methods to boost the results? Do you see any significant change? Why or why not? 

Next, use data reduction method you have learned in class to reduce the size of data, and agian try above models. Do you get better results? Justify your answer. 

<font color = 'red'>Due date for full credit: April 4, 11:59 PM
    <br>
    Due date for partial credit: April 6, 11:59 PM.
    <br> No submission will be accepted after April 6. 
    <br> Please note that your term paper is also due April 6. 
    </font>

### Import required packages

In [3]:
import pandas as pd
import numpy as np

### Loading dataset

In [4]:
data = pd.read_csv('cardiac_arrhythmia.csv',header=None)

In [5]:
data

0    1    2    3    4    5    6    7    8    9   ...   270   271   272  \
0     75    0  190   80   91  193  371  174  121  -16 ...   0.0   9.0  -0.9   
1     56    1  165   64   81  174  401  149   39   25 ...   0.0   8.5   0.0   
2     54    0  172   95  138  163  386  185  102   96 ...   0.0   9.5  -2.4   
3     55    0  175   94  100  202  380  179  143   28 ...   0.0  12.2  -2.2   
4     75    0  190   80   88  181  360  177  103  -16 ...   0.0  13.1  -3.6   
5     13    0  169   51  100  167  321  174   91  107 ...  -0.6  12.2  -2.8   
6     40    1  160   52   77  129  377  133   77   77 ...   0.0   6.5   0.0   
7     49    1  162   54   78    0  376  157   70   67 ...   0.0   8.2  -1.9   
8     44    0  168   56   84  118  354  160   63   61 ...   0.0   7.0  -1.3   
9     50    1  167   67   89  130  383  156   73   85 ...  -0.6  10.8  -1.7   
10    62    0  170   72  102  135  401  156   83   72 ...  -0.5   9.0  -2.0   
11    45    1  165   86   77  143  373  150   65   12 ...   0.0   4.4  -2.2   
12    54    1  172   58   78  155  382  163   81  -24 ...   0.0   6.3  -2.1   
13    30    0  170   73   91  180  355  157  104   68 ...  -0.9  12.3   0.0   
14    44    1  160   88   77  158  399  163   94   46 ...  -0.6  12.4   0.0   
15    47    1  150   48   75  132  350  169   65   36 ...   0.0   7.7  -0.8   
16    47    0  171   59   82  145  347  169   61   77 ...   0.0   9.4  -1.7   
17    46    1  158   58   70  120  353  122   52   57 ...   0.0   6.6   0.0   
18    73    0  165   63   91  154  392  175   83   73 ...   0.0   5.7   0.0   
19    57    1  166   72   82  181  399  158   79  -12 ...   0.0   7.7  -0.9   
20    28    1  160   58   83  251  383  189  183   50 ...  -0.6   9.1  -1.4   
21    45    0  169   67   90  122  336  177   78   81 ...  -0.6   8.3  -1.8   
22    36    1  153   75   71  132  364  169   82   62 ...   0.0   8.9  -1.0   
23    57    1  165   59   75  157  406  143   92    4 ...   0.0   6.7  -0.5   
24    40    1  153   55   82  140  388  149   82   52 ...   0.0  13.6   0.0   
25    44    0  169   80  109  128  382  195   60  -34 ...   0.0   6.9   0.0   
26    34    0  170   73   94  186  373  224  125   90 ...   0.0  15.3  -1.1   
27    31    1  160   54   95  161  407  168   83   10 ...   0.0  12.7  -1.8   
28    56    1  164   65   90  164  420  381   99   -8 ...   0.0   5.4   0.0   
29    51    1  160   83   96  147  400  301   82  -37 ...   0.0   7.3  -3.9   
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ... ...   ...   ...   ...   
422   29    1  162   57   83  164  359  154   69   64 ...   0.0  14.1  -2.2   
423   51    0  186   95   94  203  367  171  106   -7 ...   0.0   9.6  -3.5   
424    7    0  119   21  140  157  438  226   81  -40 ...   0.0  10.0  -2.1   
425   36    0  171   93   87  150  362  177   96   44 ...   0.0  10.3  -0.8   
426   35    1  160   53   55  163  340  162  102   40 ...   0.0   8.7  -0.5   
427   58    0  160   65  133  148  417  260   92 -158 ...  -0.4   6.4  -3.5   
428   64    0  160   63   83    0  364  120   90   29 ...   0.0   6.7  -0.4   
429    8    1  130   24   77  125  358  159   70   87 ...   0.0  11.3  -2.1   
430   11    0  138   29  123  145  361  221   80  112 ...  -3.4  19.6  -4.2   
431   47    0  166   56   79  145  381  173  101   52 ...   0.0   8.5   0.0   
432   11    0  140   42   88  123  362  228   81  -18 ...   0.0  17.1  -7.1   
433   70    0  167   60   80  149  290  128   93  -67 ...   0.0   2.7  -5.4   
434   20    0  178   65   88  155  360  163   71  -22 ...  -0.5  10.2   0.0   
435   39    1  164   62   79  155  367  153   95   50 ...   0.0   9.7  -0.7   
436   32    1  164   57   77  144  340  148   82   27 ...  -0.6   9.9  -0.6   
437   35    1  155   63   87  142  391  137   88   66 ...   0.0  10.7   0.0   
438   37    0  175   82   88  146  357  179   72    1 ...  -0.4  13.5  -1.2   
439   49    1  168   66   94  170  383  152  115   92 ...   0.0   8.2  -0.7   
440   37    0  176   72   88  153  389  172   89   67 .

### Data imputations

#### Removing the instances of group which is having just two members for better cross validation results

In [6]:
data= data[(data[279] !=8)]

#### Handling missing values

In [7]:
data=data.replace('?',np.NaN)

In [8]:
data.isnull().sum()

0        0
1        0
2        0
3        0
4        0
5        0
6        0
7        0
8        0
9        0
10       8
11      22
12       1
13     375
14       1
15       0
16       0
17       0
18       0
19       0
20       0
21       0
22       0
23       0
24       0
25       0
26       0
27       0
28       0
29       0
      ... 
250      0
251      0
252      0
253      0
254      0
255      0
256      0
257      0
258      0
259      0
260      0
261      0
262      0
263      0
264      0
265      0
266      0
267      0
268      0
269      0
270      0
271      0
272      0
273      0
274      0
275      0
276      0
277      0
278      0
279      0
Length: 280, dtype: int64

In [9]:
data_group = data.groupby([279])
data_group

In [10]:
data_group.size()

279
1     245
2      44
3      15
4      15
5      13
6      25
7       3
9       9
10     50
14      4
15      5
16     22
dtype: int64

In [11]:
X= data_group.transform(lambda  grp: grp.fillna(grp.mean()))

In [12]:
X=X.loc[:, ~(X == 0).all()]

In [13]:
cols = X.columns.values
cols

array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  15,  16,  17,
        18,  20,  21,  22,  23,  24,  25,  26,  27,  28,  29,  30,  31,
        32,  33,  34,  35,  36,  37,  38,  39,  40,  41,  42,  43,  44,
        45,  46,  47,  48,  49,  50,  51,  52,  53,  54,  55,  56,  57,
        58,  59,  60,  61,  62,  63,  64,  65,  66,  68,  70,  71,  72,
        73,  74,  75,  76,  77,  78,  79,  80,  81,  82,  84,  85,  86,
        87,  88,  89,  90,  91,  92,  93,  94,  95,  96,  97,  98,  99,
       100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112,
       113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125,
       126, 127, 128, 129, 130, 133, 134, 135, 136, 137, 138, 140, 142,
       144, 146, 147, 148, 149, 150, 152, 153, 154, 155, 158, 159, 160,
       161, 162, 163, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174,
       175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187,
       188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 19

### Scaling

In [14]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
X[cols] = scaler.fit_transform(X[cols])

In [15]:
X

0    1         2         3         4         5         6    \
1    0.674699  1.0  0.088889  0.341176  0.195489  0.332061  0.610108   
2    0.650602  0.0  0.099259  0.523529  0.624060  0.311069  0.555957   
3    0.662651  0.0  0.103704  0.517647  0.338346  0.385496  0.534296   
4    0.903614  0.0  0.125926  0.435294  0.248120  0.345420  0.462094   
5    0.156627  0.0  0.094815  0.264706  0.338346  0.318702  0.321300   
6    0.481928  1.0  0.081481  0.270588  0.165414  0.246183  0.523466   
7    0.590361  1.0  0.084444  0.282353  0.172932  0.000000  0.519856   
8    0.530120  0.0  0.093333  0.294118  0.218045  0.225191  0.440433   
9    0.602410  1.0  0.091852  0.358824  0.255639  0.248092  0.545126   
10   0.746988  0.0  0.096296  0.388235  0.353383  0.257634  0.610108   
11   0.542169  1.0  0.088889  0.470588  0.165414  0.272901  0.509025   
12   0.650602  1.0  0.099259  0.305882  0.172932  0.295802  0.541516   
13   0.361446  0.0  0.096296  0.394118  0.270677  0.343511  0.444043   
14   0.530120  1.0  0.081481  0.482353  0.165414  0.301527  0.602888   
15   0.566265  1.0  0.066667  0.247059  0.150376  0.251908  0.425993   
16   0.566265  0.0  0.097778  0.311765  0.203008  0.276718  0.415162   
17   0.554217  1.0  0.078519  0.305882  0.112782  0.229008  0.436823   
18   0.879518  0.0  0.088889  0.335294  0.270677  0.293893  0.577617   
19   0.686747  1.0  0.090370  0.388235  0.203008  0.345420  0.602888   
20   0.337349  1.0  0.081481  0.305882  0.210526  0.479008  0.545126   
21   0.542169  0.0  0.094815  0.358824  0.263158  0.232824  0.375451   
22   0.433735  1.0  0.071111  0.405882  0.120301  0.251908  0.476534   
23   0.686747  1.0  0.088889  0.311765  0.150376  0.299618  0.628159   
24   0.481928  1.0  0.071111  0.288235  0.203008  0.267176  0.563177   
25   0.530120  0.0  0.094815  0.435294  0.406015  0.244275  0.541516   
26   0.409639  0.0  0.096296  0.394118  0.293233  0.354962  0.509025   
27   0.373494  1.0  0.081481  0.282353  0.300752  0.307252  0.631769   
28   0.674699  1.0  0.087407  0.347059  0.263158  0.312977  0.678700   
29   0.614458  1.0  0.081481  0.452941  0.308271  0.280534  0.606498   
30   0.638554  0.0  0.103704  0.464706  0.225564  0.299618  0.635379   
..        ...  ...       ...       ...       ...       ...       ...   
422  0.349398  1.0  0.084444  0.300000  0.210526  0.312977  0.458484   
423  0.614458  0.0  0.120000  0.523529  0.293233  0.387405  0.487365   
424  0.084337  0.0  0.020741  0.088235  0.639098  0.299618  0.743682   
425  0.433735  0.0  0.097778  0.511765  0.240602  0.286260  0.469314   
426  0.421687  1.0  0.081481  0.276471  0.000000  0.311069  0.389892   
427  0.698795  0.0  0.081481  0.347059  0.586466  0.282443  0.667870   
428  0.771084  0.0  0.081481  0.335294  0.210526  0.000000  0.476534   
429  0.096386  1.0  0.037037  0.105882  0.165414  0.238550  0.454874   
430  0.132530  0.0  0.048889  0.135294  0.511278  0.276718  0.465704   
431  0.566265  0.0  0.090370  0.294118  0.180451  0.276718  0.537906   
432  0.132530  0.0  0.051852  0.211765  0.248120  0.234733  0.469314   
433  0.843373  0.0  0.091852  0.317647  0.187970  0.284351  0.209386   
434  0.240964  0.0  0.108148  0.347059  0.248120  0.295802  0.462094   
435  0.469880  1.0  0.087407  0.329412  0.180451  0.295802  0.487365   
436  0.385542  1.0  0.087407  0.300000  0.165414  0.274809  0.389892   
437  0.421687  1.0  0.074074  0.335294  0.240602  0.270992  0.574007   
438  0.445783  0.0  0.103704  0.447059  0.248120  0.278626  0.451264   
439  0.590361  1.0  0.093333  0.352941  0.293233  0.324427  0.545126   
440  0.445783  0.0  0.105185  0.388235  0.248120  0.291985  0.566787   
441  0.445783  1.0  0.081481  0.258824  0.142857  0.272901  0.512635   
442  0.783133  1.0  0.081481  0.258824  0.225564  0.272901  0.472924   
443  0.493976  1.0  0.072593  0.405882  0.248120  0.299618  0.548736   
444  0.349398  0.0  0.090370  0.335294  0.195489  0.272901  0.335740   
445  0.542169  0.0  0.103704  0.405882  0.270677  

In [16]:
X1 = X.values

In [17]:
X1[:,0].min()

0.0

#### Setting Target variable y

In [18]:
y = data[279]

### Splitting

In [19]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2,random_state=0)


Evaluation Strategy: We have used R square as the evaluation strategy for our models. This is because it depends on teh dataset to decide whether to use the R square or the precision recall scores. After checking all the scores coming on a logistic regression, we came to the conclusion of using the standard r square for our evaluations.

Best Model:
From the below regressions, we found the following useful metrics to decide the results.
The order of first three best models is: Linear SVM, Random Forest, Kernelized SVM. 
This is based on the r square values of the models. Linear SVM explains 65.8% of variance. 65.5% of variance is being explained by Random Forest. Kernelized SVM explains 64.6% of variance. 

## KNN

In [20]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV
knnreg = KNeighborsClassifier()
knn_grid = {'n_neighbors':[1,2, 3,5, 7, 15, 55]}
grid_search_knn = GridSearchCV(knnreg, knn_grid, cv = 3)
grid_search_knn.fit(X_train, y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform'),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'n_neighbors': [1, 2, 3, 5, 7, 15, 55]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [21]:
print('Best parameter {}'.format(grid_search_knn.best_params_))
print('Best score {}'.format(grid_search_knn.best_score_))

Best parameter {'n_neighbors': 3}
Best score 0.5972222222222222


In [22]:
knnreg = KNeighborsClassifier(n_neighbors = 3).fit(X_train, y_train)

print(knnreg.predict(X_test))
print('KNN R-squared test score: {}'
     .format(knnreg.score(X_test, y_test)))
print('KNN R-squared score (training): {:.9f}'
     .format(knnreg.score(X_train, y_train)))

[ 1  6  1  1  4  1  1  1  1  1  3  1  1  1  1  1  1 10  1  3  1  1  1  1  1
  1  1  1  3  1  1  1  1  2  1  1  6  1  1  1  1  1  1  1  1  1  1  1  1  1
  1  1  1  1  1  1  1  1  1 10  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1]
KNN R-squared test score: 0.6444444444444445
KNN R-squared score (training): 0.702777778


In [23]:
from sklearn.model_selection import cross_val_score
scores_knnreg_train = cross_val_score(knnreg, X_train, y_train, cv = 3)
print(scores_knnreg_train)
scores_knnreg_test = cross_val_score(knnreg, X_test, y_test, cv = 3)
print(scores_knnreg_test)

[ 0.55645161  0.63333333  0.60344828]
[ 0.54545455  0.5862069   0.57142857]


/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py:605: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


In [24]:
print("Average cross-validation score for training set: {:.9f}".format(scores_knnreg_train.mean()))
print("Average cross-validation score for testing set: {:.9f}".format(scores_knnreg_test.mean()))

Average cross-validation score for training set: 0.597744407
Average cross-validation score for testing set: 0.567696671


## Logistic Regression

In [25]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
logr= LogisticRegression(random_state=0)
logr_grid = {'C':[0.01, 0.1, 1, 10, 100]}
grid_search = GridSearchCV(logr, logr_grid, cv = 3)
grid_search.fit(X_train, y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=0, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'C': [0.01, 0.1, 1, 10, 100]}, pre_dispatch='2*n_jobs',
       refit=True, return_train_score='warn', scoring=None, verbose=0)

In [26]:
print('Best parameter {}'.format(grid_search.best_params_))
print('Best score {}'.format(grid_search.best_score_))

Best parameter {'C': 1}
Best score 0.6972222222222222


In [27]:
lr = LogisticRegression(C=1).fit(X_train, y_train)
print('Logistic regression R-squared test score: {}'
     .format(lr.score(X_test, y_test)))
print('Logistic regression R-squared score (training): {:.9f}'
     .format(lr.score(X_train, y_train)))

Logistic regression R-squared test score: 0.7111111111111111
Logistic regression R-squared score (training): 0.797222222


In [28]:
lr_predicted = lr.predict(X_test)
from sklearn.metrics import confusion_matrix
confusion = confusion_matrix(y_test, lr_predicted)

print('Logistic regression classifier (default settings)\n', confusion)

Logistic regression classifier (default settings)
 [[51  0  0  0  0  0  0  1  0]
 [ 6  4  0  0  0  0  0  0  1]
 [ 0  0  3  0  0  0  0  0  0]
 [ 2  0  0  1  0  0  0  1  0]
 [ 2  0  0  0  0  0  0  2  0]
 [ 5  0  0  0  0  0  0  1  0]
 [ 0  0  0  0  0  0  1  0  0]
 [ 2  0  0  0  0  0  0  4  0]
 [ 2  0  0  0  0  0  0  1  0]]


In [29]:
from sklearn.metrics import classification_report

print('Logistic regression\n',classification_report(y_test, lr_predicted, target_names=['1', '2','3','4','5','6','7','9','10','11','12','13','14','15','16']))

Logistic regression
              precision    recall  f1-score   support

          1       0.73      0.98      0.84        52
          2       1.00      0.36      0.53        11
          3       1.00      1.00      1.00         3
          4       1.00      0.25      0.40         4
          5       0.00      0.00      0.00         4
          6       0.00      0.00      0.00         6
          7       1.00      1.00      1.00         1
          9       0.40      0.67      0.50         6
         10       0.00      0.00      0.00         3

avg / total       0.66      0.71      0.64        90



/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1428: UserWarning: labels size, 9, does not match size of target_names, 15
  .format(len(labels), len(target_names))
/anaconda3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1135: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [30]:
from sklearn.model_selection import cross_val_score
scores_lr_train = cross_val_score(lr, X_train, y_train, cv = 3)
print(scores_lr_train)
scores_lr_test = cross_val_score(lr, X_test, y_test, cv = 3)
print(scores_lr_test)

[ 0.7016129   0.69166667  0.69827586]
[ 0.57575758  0.65517241  0.60714286]


/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py:605: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


In [31]:
print("Average cross-validation score for training set: {:.9f}".format(scores_lr_train.mean()))
print("Average cross-validation score for testing set: {:.9f}".format(scores_lr_test.mean()))

Average cross-validation score for training set: 0.697185144
Average cross-validation score for testing set: 0.612690949


### SVC Linear

In [32]:
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
param_grid = {'C': [0.01, 0.1, 1, 10,100]}
svc = SVC(kernel='linear',random_state=0)
grid_search = GridSearchCV(svc, param_grid,cv=3)
grid_search.fit(X_train, y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='linear',
  max_iter=-1, probability=False, random_state=0, shrinking=True,
  tol=0.001, verbose=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'C': [0.01, 0.1, 1, 10, 100]}, pre_dispatch='2*n_jobs',
       refit=True, return_train_score='warn', scoring=None, verbose=0)

In [33]:
print('Best parameter {}'.format(grid_search.best_params_))
print('Best score {}'.format(grid_search.best_score_))

Best parameter {'C': 1}
Best score 0.7


In [34]:
clf = SVC(kernel = 'linear', C=1).fit(X_train, y_train)
print('Linear SVC R-squared score (training): {:.2f}'
     .format(clf.score(X_train, y_train)))
print('Linear SVC R-squared score (test): {:.2f}'
     .format(clf.score(X_test, y_test)))

Linear SVC R-squared score (training): 0.86
Linear SVC R-squared score (test): 0.70


In [35]:
from sklearn.model_selection import cross_val_score
scores_svc_train = cross_val_score(clf, X_train, y_train, cv = 3)
print(scores_svc_train)
scores_svc_test = cross_val_score(clf, X_test, y_test, cv = 3)
print(scores_svc_test)

[ 0.65322581  0.71666667  0.73275862]
[ 0.60606061  0.68965517  0.67857143]


/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py:605: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


In [36]:
print("Average cross-validation score for training set: {:.9f}".format(scores_svc_train.mean()))
print("Average cross-validation score for testing set: {:.9f}".format(scores_svc_test.mean()))

Average cross-validation score for training set: 0.700883698
Average cross-validation score for testing set: 0.658095736


### SVC-Kernel

In [37]:
svc = SVC(kernel = 'rbf',random_state=0)
param_grid_kernel = {'C': [ 0.01, 0.1, 1, 10,100], 
            'gamma': [0.001,0.01, 0.1, 1]}

In [38]:
grid_search_kernel = GridSearchCV(svc, param_grid_kernel, cv = 3)
grid_search_kernel.fit(X_train, y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=0, shrinking=True,
  tol=0.001, verbose=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'C': [0.01, 0.1, 1, 10, 100], 'gamma': [0.001, 0.01, 0.1, 1]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [39]:
print('Best parameter {}'.format(grid_search_kernel.best_params_))
print('Best score {}'.format(grid_search_kernel.best_score_))

Best parameter {'C': 100, 'gamma': 0.01}
Best score 0.7083333333333334


In [40]:
clf=SVC(kernel = 'rbf',C=100,gamma=.01,random_state=0)
svc_K=clf.fit(X_train, y_train)
print('Kernelized SVC R-squared score (training): {:.9f}'
     .format(svc_K.score(X_train, y_train)))
print('Kernelized SVC R-squared score (test): {:.9f}'
     .format(svc_K.score(X_test, y_test)))

Kernelized SVC R-squared score (training): 0.930555556
Kernelized SVC R-squared score (test): 0.711111111


In [41]:
scores_clf_train = cross_val_score(svc_K, X_train, y_train, cv = 3)
print(scores_clf_train)
scores_clf_test = cross_val_score(svc_K, X_test, y_test, cv = 3)
print(scores_clf_test)

[ 0.68548387  0.71666667  0.72413793]
[ 0.60606061  0.68965517  0.64285714]


/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py:605: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


In [42]:
print("Average cross-validation score for training set: {:.9f}".format(scores_clf_train.mean()))
print("Average cross-validation score for testing set: {:.9f}".format(scores_clf_test.mean()))

Average cross-validation score for training set: 0.708762823
Average cross-validation score for testing set: 0.646190974


### Decision Tree

In [43]:
from sklearn.tree import DecisionTreeClassifier
Decision_tree=DecisionTreeClassifier(random_state=0)
param_grid = {'max_depth': [ 1,3,5,7,10,15]}
grid_search = GridSearchCV(Decision_tree, param_grid, cv = 3)
grid_search.fit(X_train, y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=0,
            splitter='best'),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'max_depth': [1, 3, 5, 7, 10, 15]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [44]:
print('Best parameter {}'.format(grid_search.best_params_))
print('Best score {}'.format(grid_search.best_score_))

Best parameter {'max_depth': 3}
Best score 0.6222222222222222


In [45]:
clf2 = DecisionTreeClassifier(max_depth = 3).fit(X_train, y_train)

print('Accuracy of Decision Tree classifier on training set: {:.9f}'
     .format(clf2.score(X_train, y_train)))
print('Accuracy of Decision Tree classifier on test set: {:.9f}'
     .format(clf2.score(X_test, y_test)))

Accuracy of Decision Tree classifier on training set: 0.677777778
Accuracy of Decision Tree classifier on test set: 0.677777778


In [46]:
scores_clf2_train = cross_val_score(clf2, X_train, y_train, cv = 3)
print(scores_clf2_train)
scores_clf2_test = cross_val_score(clf2, X_test, y_test, cv = 3)
print(scores_clf2_test)

[ 0.58064516  0.59166667  0.68965517]
[ 0.57575758  0.48275862  0.64285714]


/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py:605: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


In [47]:
print("Average cross-validation score for training set: {:.9f}".format(scores_clf2_train.mean()))
print("Average cross-validation score for testing set: {:.9f}".format(scores_clf2_test.mean()))

Average cross-validation score for training set: 0.620655667
Average cross-validation score for testing set: 0.567124446


### Ensembles of Decision Trees: Random Forrest

In [48]:
from sklearn.ensemble import RandomForestClassifier
clf3 = RandomForestClassifier(random_state=0)

In [49]:
param_grid = {'max_features': [ 1,3,5,8,10,15,20,30]}
grid_search = GridSearchCV(clf3, param_grid, cv = 3)
grid_search.fit(X_train, y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=0, verbose=0, warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'max_features': [1, 3, 5, 8, 10, 15, 20, 30]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [50]:
print('Best parameter {}'.format(grid_search.best_params_))
print('Best score {}'.format(grid_search.best_score_))

Best parameter {'max_features': 30}
Best score 0.6777777777777778


In [51]:
clf3 = RandomForestClassifier(max_features = 10, random_state = 0)
Random_forest=clf3.fit(X_train, y_train)

print(Random_forest.predict(X_test))
print('Random Forest R-squared test score: {}'
     .format(Random_forest.score(X_test, y_test)))
print('Random Forest R-squared score (training): {:.9f}'
     .format(Random_forest.score(X_train, y_train)))

[ 1  1  1  2  2  1  1  1  2  1  3  1  1  1  1  1  1 10  1  3  1  1  2  1  2
  1 10 10  3  5  1 10  1  1  1  2  1 10  1  1  1  1  1  1  1  1  1  1  1  1
  1  1  1  1  2  1  1 10  1 10 10  1  2  1  9  1  1  1  1  1  1  2  1  1 10
  1  1  1  2  1 10  1  1  1  1  1  1  6  1  1]
Random Forest R-squared test score: 0.7
Random Forest R-squared score (training): 0.983333333


In [52]:
from sklearn.model_selection import cross_val_score
scores_Random_forest_train = cross_val_score(Random_forest, X_train, y_train, cv = 3)
print(scores_Random_forest_train)
scores_Random_forest_test = cross_val_score(Random_forest, X_test, y_test, cv = 3)
print(scores_Random_forest_test)

[ 0.63709677  0.675       0.64655172]
[ 0.66666667  0.62068966  0.67857143]


/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py:605: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


In [53]:
print("Average cross-validation score for training set: {:.9f}".format(scores_Random_forest_train.mean()))
print("Average cross-validation score for testing set: {:.9f}".format(scores_Random_forest_test.mean()))

Average cross-validation score for training set: 0.652882833
Average cross-validation score for testing set: 0.655309250


### Gradient-boosted decision trees

In [54]:
from sklearn.ensemble import GradientBoostingClassifier
clf4 = GradientBoostingClassifier(random_state=0)

In [55]:
param_grid = {'learning_rate': [ 0.01,0.1,1,10,20],'max_depth': [ 1,2,3,5,8,10,15]}
grid_search = GridSearchCV(clf4, param_grid, cv = 3)
grid_search.fit(X_train, y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=GradientBoostingClassifier(criterion='friedman_mse', init=None,
              learning_rate=0.1, loss='deviance', max_depth=3,
              max_features=None, max_leaf_nodes=None,
              min_impurity_decrease=0.0, min_impurity_split=None,
              min_samples_leaf=1, min_samples_split=2,
              min_weight_fraction_leaf=0.0, n_estimators=100,
              presort='auto', random_state=0, subsample=1.0, verbose=0,
              warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'learning_rate': [0.01, 0.1, 1, 10, 20], 'max_depth': [1, 2, 3, 5, 8, 10, 15]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [56]:
print('Best parameter {}'.format(grid_search.best_params_))
print('Best score {}'.format(grid_search.best_score_))

Best parameter {'learning_rate': 0.01, 'max_depth': 1}
Best score 0.6944444444444444


In [57]:
clf = GradientBoostingClassifier(learning_rate = 0.1, max_depth = 1, random_state = 0)
clf.fit(X_train, y_train)

print('Accuracy of GBDT classifier on training set: {:.2f}'
     .format(clf.score(X_train, y_train)))
print('Accuracy of GBDT classifier on test set: {:.2f}'
     .format(clf.score(X_test, y_test)))

Accuracy of GBDT classifier on training set: 0.94
Accuracy of GBDT classifier on test set: 0.71


In [58]:
from sklearn.model_selection import cross_val_score
scores_GBDT_train = cross_val_score(clf, X_train, y_train, cv = 3)
print(scores_GBDT_train)
scores_GBDT_test = cross_val_score(clf, X_test, y_test, cv = 3)
print(scores_GBDT_test)

[ 0.62903226  0.74166667  0.71551724]


/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py:605: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


[ 0.6969697   0.65517241  0.75      ]


In [59]:
print("Average cross-validation score for training set: {:.9f}".format(scores_GBDT_train.mean()))
print("Average cross-validation score for testing set: {:.9f}".format(scores_GBDT_test.mean()))

Average cross-validation score for training set: 0.695405389
Average cross-validation score for testing set: 0.700714037


### Voting Classifier
The following code creates and train a voting classifier in Scikit Learn composed of following diverse classifiers
1. KNN classifcation
2. Logistic Regression
3. Support Vector Machine
4. Decision Tree
5. Random Forest

### Hard Voting

In [60]:
from sklearn.ensemble import VotingClassifier

knn_clf=KNeighborsClassifier()
log_clf = LogisticRegression(random_state=0)
svm_clf = SVC(random_state=0)
dt_clf=DecisionTreeClassifier(random_state=0)
rnd_clf = RandomForestClassifier(random_state=0)

voting_clf = VotingClassifier(estimators=[('knn', knn_clf),
                                          ('lr', log_clf),
                                          ('svm', svm_clf),
                                          ('rf', rnd_clf), 
                                          ('dt', dt_clf)], voting='hard')
voting_clf.fit(X_train, y_train)

VotingClassifier(estimators=[('knn', KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')), ('lr', LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1,...         min_weight_fraction_leaf=0.0, presort=False, random_state=0,
            splitter='best'))],
         flatten_transform=None, n_jobs=1, voting='hard', weights=None)

### Soft Voting

In [62]:
log_clf = LogisticRegression(random_state=0)
rnd_clf = RandomForestClassifier(random_state=0)
svm_clf = SVC(probability=True, random_state=0)
dt_clf=DecisionTreeClassifier(random_state=0)
knn_clf=KNeighborsClassifier()

voting_clf = VotingClassifier(estimators=[('knn', knn_clf),
                                          ('lr', log_clf),
                                          ('svm', svm_clf),
                                          ('rf', rnd_clf), 
                                          ('dt', dt_clf)], voting='soft')
voting_clf.fit(X_train, y_train)

VotingClassifier(estimators=[('knn', KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')), ('lr', LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1,...         min_weight_fraction_leaf=0.0, presort=False, random_state=0,
            splitter='best'))],
         flatten_transform=None, n_jobs=1, voting='soft', weights=None)

In [63]:
from sklearn.metrics import accuracy_score

for clf in (knn_clf, log_clf, rnd_clf, svm_clf, dt_clf, voting_clf):
    clf.fit(X_train, y_train)
    y_pred = clf.predict(X_test)
    print(clf.__class__.__name__, accuracy_score(y_test, y_pred))

KNeighborsClassifier 0.622222222222
LogisticRegression 0.711111111111
RandomForestClassifier 0.7
SVC 0.577777777778
DecisionTreeClassifier 0.633333333333
VotingClassifier 0.722222222222


## Bagging Ensembles

Implemtation of different bagging and boosting techniques.
Following results have been interpreted from the below regressions:

Bagging: Well, there has been significant change in the performance of the model if we have used bagging instead of no bagging.
    The first block of code shows use of bagging classifier, while the second block shows the results without the bagging classifiers. 
    The order of best 3 models after using bagging are: Random Forest, decision trees and Logistic Regression. 
    
    implementing random forest gave r square result of 73.3%, which is the highest percentage of variance explained by any model in our project. The main reason for this could be that random forest is an ensemble approach to the decision trees and performs bagging on all the data each time.
    
    Decision tree and logistic regression explain 72% of variance in the data. 
    
    Also, We see no difference in the accuracy of the model if we use bagging in logistic regression because logistic regression is already a low variance estimator.
    
 Boosting: Adaptive boosting has been applied on all the models except KNN. Gradient boosting has limited applicabiltiy.
 Results: The order of models which show best results are: decision tree, linear SVM and random forrest. The boosting models in general do not show good results in terms of r square values among regular method, bagging and PCA. The highest r square we could get is 71% from decision tree.
 

#### 1.Decision Trees

In [64]:
from sklearn.ensemble import BaggingClassifier
bag_clf = BaggingClassifier(DecisionTreeClassifier(random_state=0), 
                            n_estimators=500, max_samples=100, bootstrap=True, n_jobs=-1, random_state=0)

bag_clf.fit(X_train, y_train)
y_pred = bag_clf.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.722222222222


In [65]:
tree_clf = DecisionTreeClassifier(random_state=0)
tree_clf.fit(X_train, y_train)
y_pred_tree = tree_clf.predict(X_test)
print(accuracy_score(y_test, y_pred_tree))

0.633333333333


#### 2. Random Forests

In [66]:
bag_clf = BaggingClassifier(DecisionTreeClassifier(splitter="random", max_leaf_nodes=16, random_state=0),
                            n_estimators=500, max_samples=1.0, bootstrap=True, n_jobs=-1, random_state=0)

bag_clf.fit(X_train, y_train)
y_pred = bag_clf.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.733333333333


In [67]:
rnd_clf = RandomForestClassifier(n_estimators=500, max_leaf_nodes=16, n_jobs=-1, random_state=0)
rnd_clf.fit(X_train, y_train)
y_pred_rf = rnd_clf.predict(X_test)
print(accuracy_score(y_test, y_pred_rf))

0.711111111111


In [68]:
np.sum(y_pred == y_pred_rf) / len(y_pred)

0.96666666666666667

#### 3. KNN

In [69]:
bag_clf = BaggingClassifier(KNeighborsClassifier(n_neighbors = 3),
                            n_estimators=500, max_samples=1.0, bootstrap=True, n_jobs=-1, random_state=0)

bag_clf.fit(X_train, y_train)
y_pred = bag_clf.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.666666666667


In [70]:
knn_clf = KNeighborsClassifier(n_neighbors = 3, n_jobs=-1)
knn_clf.fit(X_train, y_train)
y_pred_knn = knn_clf.predict(X_test)
print(accuracy_score(y_test, y_pred_knn))

0.644444444444


#### 4. Logistic Regression

We see no difference in the accuracy of the model if we use bagging in logistic regression because logistic regression is already a low variance estimator

In [71]:
bag_clf = BaggingClassifier(LogisticRegression(random_state=0),
                            n_estimators=500, max_samples=1.0, bootstrap=True, n_jobs=-1, random_state=0)

bag_clf.fit(X_train, y_train)
y_pred = bag_clf.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.711111111111


In [72]:
lr_clf = LogisticRegression(random_state=0)
lr_clf.fit(X_train, y_train)
y_pred_lr = lr_clf.predict(X_test)
print(accuracy_score(y_test, y_pred_lr))

0.711111111111


#### 5. SVC

In [73]:
bag_clf = BaggingClassifier(SVC(kernel='linear',random_state=0),
                            n_estimators=500, max_samples=1.0, bootstrap=True, n_jobs=-1, random_state=0)

bag_clf.fit(X_train, y_train)
y_pred = bag_clf.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.711111111111


In [74]:
clf = SVC(kernel='linear',random_state=0)
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.7


#### 6. SVC- Kernel

There is no effect of bagging on kernalized SVC

In [75]:
bag_clf = BaggingClassifier(SVC(kernel = 'rbf',random_state=0),
                            n_estimators=500, max_samples=1.0, bootstrap=True, n_jobs=-1, random_state=0)

bag_clf.fit(X_train, y_train)
y_pred = bag_clf.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.577777777778


In [76]:
clf = SVC(kernel='rbf',random_state=0)
clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.577777777778


#### Out-of-Bag evaluation

In [102]:
bag_clf = BaggingClassifier(DecisionTreeClassifier(random_state=0), n_estimators=500,
                             bootstrap=True,n_jobs=-1, oob_score=True, random_state=0)
bag_clf.fit(X_train, y_train)
bag_clf.oob_score_

0.7055555555555556

In [103]:
y_pred = bag_clf.predict(X_test)
accuracy_score(y_test, y_pred)

0.73333333333333328

## Boosting

### 1. AdaBoost

#### 1. Decision Tree

In [97]:
from sklearn.ensemble import AdaBoostClassifier

ada_clf = AdaBoostClassifier(DecisionTreeClassifier(max_depth=3), n_estimators=200,
                             algorithm="SAMME", learning_rate=0.5, random_state=0)
ada_clf.fit(X_train, y_train)
y_pred = ada_clf.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.711111111111


#### 2. Random Forest

In [101]:
ada_clf = AdaBoostClassifier(RandomForestClassifier(max_features = 10, random_state = 0), n_estimators=200,
                             algorithm="SAMME.R", learning_rate=0.5, random_state=0)
ada_clf.fit(X_train, y_train)
y_pred = ada_clf.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.644444444444


#### 3. Logistic Regression

In [100]:
ada_clf = AdaBoostClassifier(LogisticRegression(C=1, random_state=0), n_estimators=200,
                             algorithm="SAMME.R", learning_rate=0.5, random_state=0)
ada_clf.fit(X_train, y_train)
y_pred = ada_clf.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.655555555556


#### 5. SVM

In [82]:
ada_clf = AdaBoostClassifier(SVC(kernel='linear', C=1,random_state=0), n_estimators=200,
                             algorithm="SAMME", learning_rate=0.5, random_state=0)
ada_clf.fit(X_train, y_train)
y_pred = ada_clf.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.577777777778


#### 6. SVM-Kernel

In [83]:
ada_clf = AdaBoostClassifier(SVC(kernel = 'rbf',C=100,gamma=.01,random_state=0), n_estimators=200,
                             algorithm="SAMME", learning_rate=0.5, random_state=0)
ada_clf.fit(X_train, y_train)
y_pred = ada_clf.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.577777777778


###  2. Gradient Boosting

In [84]:
from sklearn.ensemble import GradientBoostingClassifier

gbrt = GradientBoostingClassifier(max_depth=2, n_estimators=3, learning_rate=1.0, random_state=0)
gbrt.fit(X, y)

GradientBoostingClassifier(criterion='friedman_mse', init=None,
              learning_rate=1.0, loss='deviance', max_depth=2,
              max_features=None, max_leaf_nodes=None,
              min_impurity_decrease=0.0, min_impurity_split=None,
              min_samples_leaf=1, min_samples_split=2,
              min_weight_fraction_leaf=0.0, n_estimators=3, presort='auto',
              random_state=0, subsample=1.0, verbose=0, warm_start=False)

In [85]:
print('Accuracy of GradientBoosting classifier on training set: {:.9f}'
     .format(gbrt.score(X_train, y_train)))
print('Accuracy of GradientBoosting classifier on test set: {:.9f}'
     .format(gbrt.score(X_test, y_test)))

Accuracy of GradientBoosting classifier on training set: 0.091666667
Accuracy of GradientBoosting classifier on test set: 0.122222222


In [86]:
gbrt_grid = GradientBoostingClassifier()

In [88]:
param_grid = {'max_depth': [ 1,2,3,5,8,10],'n_estimators':[1,2,3,5,8,10],'learning_rate':[0.01,0.1,1,10]}
grid_search = GridSearchCV(gbrt_grid, param_grid, cv = 3)
grid_search.fit(X_train, y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=GradientBoostingClassifier(criterion='friedman_mse', init=None,
              learning_rate=0.1, loss='deviance', max_depth=3,
              max_features=None, max_leaf_nodes=None,
              min_impurity_decrease=0.0, min_impurity_split=None,
              min_samples_leaf=1, min_samples_split=2,
              min_weight_fraction_leaf=0.0, n_estimators=100,
              presort='auto', random_state=None, subsample=1.0, verbose=0,
              warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'max_depth': [1, 2, 3, 5, 8, 10], 'n_estimators': [1, 2, 3, 5, 8, 10], 'learning_rate': [0.01, 0.1, 1, 10]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [89]:
print('Best parameter {}'.format(grid_search.best_params_))
print('Best score {}'.format(grid_search.best_score_))

Best parameter {'learning_rate': 0.1, 'max_depth': 1, 'n_estimators': 10}
Best score 0.6944444444444444


In [90]:
gbrt_best = GradientBoostingClassifier(max_depth=1, n_estimators=10, learning_rate=0.1, random_state=0)
gbrt_best.fit(X, y)

GradientBoostingClassifier(criterion='friedman_mse', init=None,
              learning_rate=0.1, loss='deviance', max_depth=1,
              max_features=None, max_leaf_nodes=None,
              min_impurity_decrease=0.0, min_impurity_split=None,
              min_samples_leaf=1, min_samples_split=2,
              min_weight_fraction_leaf=0.0, n_estimators=10,
              presort='auto', random_state=0, subsample=1.0, verbose=0,
              warm_start=False)

In [91]:
print('Accuracy of AdaBoost classifier on training set: {:.9f}'
     .format(gbrt_best.score(X_train, y_train)))
print('Accuracy of AdaBoost classifier on test set: {:.9f}'
     .format(gbrt_best.score(X_test, y_test)))

Accuracy of AdaBoost classifier on training set: 0.744444444
Accuracy of AdaBoost classifier on test set: 0.788888889


In [295]:
gbrt_slow = GradientBoostingClassifier(max_depth=2, n_estimators=200, learning_rate=0.1, random_state=0)
gbrt_slow.fit(X, y)

GradientBoostingClassifier(criterion='friedman_mse', init=None,
              learning_rate=0.1, loss='deviance', max_depth=2,
              max_features=None, max_leaf_nodes=None,
              min_impurity_decrease=0.0, min_impurity_split=None,
              min_samples_leaf=1, min_samples_split=2,
              min_weight_fraction_leaf=0.0, n_estimators=200,
              presort='auto', random_state=0, subsample=1.0, verbose=0,
              warm_start=False)

In [296]:
print('Accuracy of AdaBoost classifier on training set: {:.9f}'
     .format(gbrt_slow.score(X_train, y_train)))
print('Accuracy of AdaBoost classifier on test set: {:.9f}'
     .format(gbrt_slow.score(X_test, y_test)))

Accuracy of AdaBoost classifier on training set: 1.000000000
Accuracy of AdaBoost classifier on test set: 1.000000000


## Dimensionality Reduction

### PCA

In [107]:
from sklearn.decomposition import PCA
pca = PCA(n_components = 2)
X2D = pca.fit_transform(X)

Recover the 3D points projected on the plane (PCA 2D subspace)

In [108]:
X3D_inv = pca.inverse_transform(X2D)

In [109]:
np.mean(np.sum(np.square(X3D_inv - X), axis=1))

3.0773156705730376

In [110]:
pca.components_

array([[  6.97511808e-02,  -6.12411694e-02,   2.16945171e-03,
          1.28833960e-02,   2.04167428e-02,  -1.53924894e-03,
          4.29224954e-03,   2.45989033e-02,   1.58684575e-02,
         -1.47211540e-01,   2.08248065e-02,   3.57916388e-02,
         -5.34034646e-02,  -8.64249796e-03,   2.77052508e-02,
         -2.04342109e-03,   2.65535172e-03,   1.89314278e-02,
          7.41905121e-03,   1.10929997e-02,  -3.67769925e-04,
         -7.25215124e-02,  -7.63293624e-02,   2.14892199e-01,
          1.15491668e-02,   1.96690872e-03,  -7.03410859e-02,
          1.50731610e-02,  -2.87614080e-02,  -8.88058868e-03,
          4.96825141e-04,  -1.60442563e-03,   1.74423911e-03,
          4.79715040e-04,  -1.85415454e-01,   1.62663022e-01,
          1.21331453e-02,   4.30696860e-03,  -1.52632398e-01,
         -2.61681547e-03,  -2.00840991e-02,  -3.04355364e-03,
         -5.34303600e-03,  -2.63019642e-03,   2.01633255e-03,
          2.45441034e-02,   1.26012642e-01,  -9.42207722e-02,
        

#### Variance ratio:
The first dimension explains 11.19% of the variance, while the second explains 9.93%.

In [111]:
pca.explained_variance_ratio_

array([ 0.11194233,  0.09399226])

By projecting down to 2D, we lost about 79.4% of the variance:

In [112]:
1 - pca.explained_variance_ratio_.sum()

0.79406541156399379

### Manifold Learning

In [113]:
from six.moves import urllib
from sklearn.datasets import fetch_mldata

In [114]:
pca = PCA()
pca.fit(X_train)
cumsum = np.cumsum(pca.explained_variance_ratio_)
d = np.argmax(cumsum >= 0.95) + 1

In [115]:
d

83

In [116]:
pca = PCA(n_components=0.95)
X_reduced = pca.fit_transform(X_train)

In [117]:
pca.n_components_

83

In [118]:
np.sum(pca.explained_variance_ratio_)

0.95127441954013714

In [123]:
pca = PCA(n_components = 83)
X_reduced = pca.fit_transform(X_train)
X_recovered = pca.inverse_transform(X_reduced)

In [130]:
X_test_reduced = pca.fit_transform(X_test)
X_recovered = pca.inverse_transform(X_test_reduced)

In [136]:
X_reduced.shape

(360, 83)

## KNN

Average cross-validation score for testing and training set is lamost same when we use PCA

In [131]:
X_train=X_reduced
X_test = X_test_reduced
knn_grid = {'n_neighbors':[1,2, 3,5, 7, 15, 55]}
grid_search_knn = GridSearchCV(knnreg, knn_grid, cv = 3)
grid_search_knn.fit(X_train, y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=3, p=2,
           weights='uniform'),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'n_neighbors': [1, 2, 3, 5, 7, 15, 55]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [132]:
print('Best parameter {}'.format(grid_search_knn.best_params_))
print('Best score {}'.format(grid_search_knn.best_score_))

Best parameter {'n_neighbors': 3}
Best score 0.6055555555555555


In [133]:
knnreg = KNeighborsClassifier(n_neighbors = 3).fit(X_train, y_train)

print(knnreg.predict(X_test))
print('KNN R-squared test score: {}'
     .format(knnreg.score(X_test, y_test)))
print('KNN R-squared score (training): {:.9f}'
     .format(knnreg.score(X_train, y_train)))

[ 2  1  1  1  1  1  1  1  1  1 10  1  1  1  1  1  1  1  1  1  2  1  6  5  6
  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  2
  1  1  1  1  1  1  1  1  1  3  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1
  1  1  1  1  2  5 10  1  1  1  1  1  2  1  1]
KNN R-squared test score: 0.5333333333333333
KNN R-squared score (training): 0.702777778


In [134]:
from sklearn.model_selection import cross_val_score
scores_knnreg_train = cross_val_score(knnreg, X_train, y_train, cv = 3)
print(scores_knnreg_train)
scores_knnreg_test = cross_val_score(knnreg, X_test, y_test, cv = 3)
print(scores_knnreg_test)

[ 0.57258065  0.63333333  0.61206897]
[ 0.54545455  0.5862069   0.57142857]


/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py:605: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


In [137]:
print("Average cross-validation score for training set: {:.9f}".format(scores_knnreg_train.mean()))
print("Average cross-validation score for testing set: {:.9f}".format(scores_knnreg_test.mean()))

Average cross-validation score for training set: 0.605994315
Average cross-validation score for testing set: 0.567696671


## Logistic Regression

Average cross-validation score for testing and training set is little more when we use PCA

In [138]:
logr= LogisticRegression(random_state=0)
logr_grid = {'C':[0.01, 0.1, 1, 10, 100]}
grid_search = GridSearchCV(logr, logr_grid, cv = 3)
grid_search.fit(X_train, y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=0, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'C': [0.01, 0.1, 1, 10, 100]}, pre_dispatch='2*n_jobs',
       refit=True, return_train_score='warn', scoring=None, verbose=0)

In [139]:
print('Best parameter {}'.format(grid_search.best_params_))
print('Best score {}'.format(grid_search.best_score_))

Best parameter {'C': 10}
Best score 0.7055555555555556


In [141]:
lr = LogisticRegression(C=10).fit(X_train, y_train)
print('Logistic regression R-squared test score: {}'
     .format(lr.score(X_test, y_test)))
print('Logistic regression R-squared score (training): {:.9f}'
     .format(lr.score(X_train, y_train)))

Logistic regression R-squared test score: 0.5111111111111111
Logistic regression R-squared score (training): 0.875000000


In [142]:
lr_predicted = lr.predict(X_test)
from sklearn.metrics import confusion_matrix
confusion = confusion_matrix(y_test, lr_predicted)

print('Logistic regression classifier (default settings)\n', confusion)

Logistic regression classifier (default settings)
 [[45  2  0  1  0  1  0  0  3  0]
 [ 8  0  0  1  0  1  0  0  1  0]
 [ 0  0  0  0  0  0  0  0  3  0]
 [ 3  1  0  0  0  0  0  0  0  0]
 [ 3  0  1  0  0  0  0  0  0  0]
 [ 4  0  0  0  0  0  1  0  1  0]
 [ 0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  1  0  0  0  0  0]
 [ 5  0  0  0  0  0  0  0  1  0]
 [ 1  1  0  0  0  0  0  0  1  0]]


In [143]:
scores_lr_train = cross_val_score(lr, X_train, y_train, cv = 3)
print(scores_lr_train)
scores_lr_test = cross_val_score(lr, X_test, y_test, cv = 3)
print(scores_lr_test)

[ 0.67741935  0.725       0.71551724]
[ 0.63636364  0.72413793  0.71428571]


/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py:605: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


In [144]:
print("Average cross-validation score for training set: {:.9f}".format(scores_lr_train.mean()))
print("Average cross-validation score for testing set: {:.9f}".format(scores_lr_test.mean()))

Average cross-validation score for training set: 0.705978865
Average cross-validation score for testing set: 0.691595761


### SVC Linear

Average cross-validation score for testing and training set is little more when we use PCA

In [145]:
param_grid = {'C': [0.01, 0.1, 1, 10,100]}
svc = SVC(kernel='linear',random_state=0)
grid_search = GridSearchCV(svc, param_grid,cv=3)
grid_search.fit(X_train, y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='linear',
  max_iter=-1, probability=False, random_state=0, shrinking=True,
  tol=0.001, verbose=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'C': [0.01, 0.1, 1, 10, 100]}, pre_dispatch='2*n_jobs',
       refit=True, return_train_score='warn', scoring=None, verbose=0)

In [146]:
print('Best parameter {}'.format(grid_search.best_params_))
print('Best score {}'.format(grid_search.best_score_))

Best parameter {'C': 1}
Best score 0.7027777777777777


In [147]:
clf = SVC(kernel = 'linear', C=1).fit(X_train, y_train)
print('Linear SVC R-squared score (training): {:.2f}'
     .format(clf.score(X_train, y_train)))
print('Linear SVC R-squared score (test): {:.2f}'
     .format(clf.score(X_test, y_test)))

Linear SVC R-squared score (training): 0.85
Linear SVC R-squared score (test): 0.51


In [148]:
scores_svc_train = cross_val_score(clf, X_train, y_train, cv = 3)
print(scores_svc_train)
scores_svc_test = cross_val_score(clf, X_test, y_test, cv = 3)
print(scores_svc_test)

[ 0.66129032  0.70833333  0.74137931]
[ 0.60606061  0.68965517  0.67857143]


/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py:605: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


In [149]:
print("Average cross-validation score for training set: {:.9f}".format(scores_svc_train.mean()))
print("Average cross-validation score for testing set: {:.9f}".format(scores_svc_test.mean()))

Average cross-validation score for training set: 0.703667655
Average cross-validation score for testing set: 0.658095736


### SVC-Kernel

Average cross-validation score for testing and training set is same for SVC-Kernel with and without PCA

In [151]:
svc = SVC(kernel = 'rbf',random_state=0)
param_grid_kernel = {'C': [ 0.01, 0.1, 1, 10,100], 
            'gamma': [0.001,0.01, 0.1, 1]}

In [152]:
grid_search_kernel = GridSearchCV(svc, param_grid_kernel, cv = 3)
grid_search_kernel.fit(X_train, y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=0, shrinking=True,
  tol=0.001, verbose=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'C': [0.01, 0.1, 1, 10, 100], 'gamma': [0.001, 0.01, 0.1, 1]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [153]:
print('Best parameter {}'.format(grid_search_kernel.best_params_))
print('Best score {}'.format(grid_search_kernel.best_score_))

Best parameter {'C': 100, 'gamma': 0.01}
Best score 0.7083333333333334


In [154]:
clf=SVC(kernel = 'rbf',C=100,gamma=.01,random_state=0)
svc_K=clf.fit(X_train, y_train)
print('Kernelized SVC R-squared score (training): {:.9f}'
     .format(svc_K.score(X_train, y_train)))
print('Kernelized SVC R-squared score (test): {:.9f}'
     .format(svc_K.score(X_test, y_test)))

Kernelized SVC R-squared score (training): 0.897222222
Kernelized SVC R-squared score (test): 0.500000000


In [155]:
scores_clf_train = cross_val_score(svc_K, X_train, y_train, cv = 3)
print(scores_clf_train)
scores_clf_test = cross_val_score(svc_K, X_test, y_test, cv = 3)
print(scores_clf_test)

[ 0.69354839  0.68333333  0.75      ]
[ 0.60606061  0.68965517  0.64285714]


/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py:605: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


In [156]:
print("Average cross-validation score for training set: {:.9f}".format(scores_clf_train.mean()))
print("Average cross-validation score for testing set: {:.9f}".format(scores_clf_test.mean()))

Average cross-validation score for training set: 0.708960573
Average cross-validation score for testing set: 0.646190974


### Decision Tree

In [157]:
Decision_tree=DecisionTreeClassifier(random_state=0)
param_grid = {'max_depth': [ 1,3,5,7,10,15]}
grid_search = GridSearchCV(Decision_tree, param_grid, cv = 3)
grid_search.fit(X_train, y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=0,
            splitter='best'),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'max_depth': [1, 3, 5, 7, 10, 15]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [158]:
print('Best parameter {}'.format(grid_search.best_params_))
print('Best score {}'.format(grid_search.best_score_))

Best parameter {'max_depth': 1}
Best score 0.5444444444444444


In [160]:
clf2 = DecisionTreeClassifier(max_depth = 3).fit(X_train, y_train)

print('Accuracy of Decision Tree classifier on training set: {:.9f}'
     .format(clf2.score(X_train, y_train)))
print('Accuracy of Decision Tree classifier on test set: {:.9f}'
     .format(clf2.score(X_test, y_test)))

Accuracy of Decision Tree classifier on training set: 0.627777778
Accuracy of Decision Tree classifier on test set: 0.566666667


In [161]:
scores_clf2_train = cross_val_score(clf2, X_train, y_train, cv = 3)
print(scores_clf2_train)
scores_clf2_test = cross_val_score(clf2, X_test, y_test, cv = 3)
print(scores_clf2_test)

[ 0.55645161  0.525       0.54310345]
[ 0.48484848  0.48275862  0.42857143]


/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py:605: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


In [162]:
print("Average cross-validation score for training set: {:.9f}".format(scores_clf2_train.mean()))
print("Average cross-validation score for testing set: {:.9f}".format(scores_clf2_test.mean()))

Average cross-validation score for training set: 0.541518354
Average cross-validation score for testing set: 0.465392845


### Ensembles of Decision Trees: Random Forrest

In [163]:
clf3 = RandomForestClassifier(random_state=0)

In [164]:
param_grid = {'max_features': [ 1,3,5,8,10,15,20,30]}
grid_search = GridSearchCV(clf3, param_grid, cv = 3)
grid_search.fit(X_train, y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=0, verbose=0, warm_start=False),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'max_features': [1, 3, 5, 8, 10, 15, 20, 30]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=0)

In [165]:
print('Best parameter {}'.format(grid_search.best_params_))
print('Best score {}'.format(grid_search.best_score_))

Best parameter {'max_features': 10}
Best score 0.575


In [166]:
clf3 = RandomForestClassifier(max_features = 10, random_state = 0)
Random_forest=clf3.fit(X_train, y_train)

print(Random_forest.predict(X_test))
print('Random Forest R-squared test score: {}'
     .format(Random_forest.score(X_test, y_test)))
print('Random Forest R-squared score (training): {:.9f}'
     .format(Random_forest.score(X_train, y_train)))

[ 1  2  1  1  1  1  1  1  1  1  3  1  1  1  1  1  1  1  1  1  1  1  3  1  1
  1  1  1  1  1  1  1  1  1 10  1  1  1  1  1  1  1  1  1  1  1 10  1  1  1
  1  1  1  1  4  1  1  1  1  3  1  1  1  1  1  1  1  1  1  1  1  1  1  1  2
  4  1  1  1  1  1  1  1  1  1  1  1 10  1  1]
Random Forest R-squared test score: 0.5555555555555556
Random Forest R-squared score (training): 0.975000000


In [167]:
from sklearn.model_selection import cross_val_score
scores_Random_forest_train = cross_val_score(Random_forest, X_train, y_train, cv = 3)
print(scores_Random_forest_train)
scores_Random_forest_test = cross_val_score(Random_forest, X_test, y_test, cv = 3)
print(scores_Random_forest_test)

[ 0.56451613  0.6         0.56034483]
[ 0.54545455  0.5862069   0.57142857]


/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_split.py:605: Warning: The least populated class in y has only 1 members, which is too few. The minimum number of members in any class cannot be less than n_splits=3.
  % (min_groups, self.n_splits)), Warning)


In [168]:
print("Average cross-validation score for training set: {:.9f}".format(scores_Random_forest_train.mean()))
print("Average cross-validation score for testing set: {:.9f}".format(scores_Random_forest_test.mean()))

Average cross-validation score for training set: 0.574953652
Average cross-validation score for testing set: 0.567696671


In dimensionality reduction, we have implemented PCA and manifold learning. We have reduced the variables to 83 components and have used all the models again. Based on the above regressions, we found following results:
Logistic Regression, KNN, and linear SVM show the best results of 69%, 56.7% and 65% r squared values. We can say that the results werent too different from the original models where we did not reduce the number of components. This could be due to the fact that there was very less correlation in the data already.  

Conclusion:
Bagging can be considered as a good technique in determining the results. We can say this based on the r squared values obtained for each model. 